include/mbedtls/rsa.h

```diff
@@ -74,7 +75,171 @@ extern "C" {
 #endif
 
 /**
- * \brief          RSA context structure
+ * Helper functions for RSA-related operations on MPI's.
+ */
+
+/**
+ * \brief          Compute RSA prime moduli P, Q from public modulus N=PQ
+ *                 and a pair of private and public key.
+ *
+ * \note           This is a 'static' helper function not operating on
+ *                 an RSA context. Alternative implementations need not
+ *                 overwrite it.
+ *
+ * \param N        RSA modulus N = PQ, with P, Q to be found
+ * \param D        RSA private exponent
+ * \param E        RSA public exponent
+ * \param f_rng    PRNG to be used for randomization, or NULL
+ * \param p_rng    PRNG context for f_rng, or NULL
+ * \param P        Pointer to MPI holding first prime factor of N on success
+ * \param Q        Pointer to MPI holding second prime factor of N on success
+ *
+ * \return
+ *                 - 0 if successful. In this case, P and Q constitute a
+ *                   factorization of N.
+ *                 - A non-zero error code otherwise.
+ *
+ * \note           It is neither checked that P, Q are prime nor that
+ *                 D, E are modular inverses wrt. P-1 and Q-1. For that,
+ *                 use the helper function \c mbedtls_rsa_validate_params.
+ *
+ */
+int mbedtls_rsa_deduce_moduli( mbedtls_mpi const *N, mbedtls_mpi const *D,
```

Comment:
I want them to be part of the public API because they will be useful for users providing alternative implementations using fewer context variables, not only for us.

Reply:
I agree that it is possible to construct an alternative RSA implementation where these are useful, but I mostly can imagine it if some new high level optimisation trick comes around, in which case we probably want to implement it ourselves. I don't see how these can be useful in our current typical use cases:
- using an HSM (MBEDTLS_PK_RSA_ALT_SUPPORT)
- using a hardware accelerator (we don't support this yet)
- using special bignum support instructions (we partially support this in `bn_mul.h`)
- using different algorithms for big integer arithmetic (we don't support this yet)

Please appreciate that every non-trivial function we add to the public API comes with a cost:
- it makes the interface more complex, therefore harder to use and
- we have to keep them for the foreseeable future or go through the pain of deprecating them 

Having said that, I don't feel that the use case for these functions is common enough to justify the cost.

Reply:
I don't agree with your evaluation of usefulness - these functions are already useful now, and they will always be needed when a user presents a key as `(N,D,E)`-only (see e.g. the recent https://github.com/ARMmbed/mbedtls/issues/1114) while the implementation needs more, e.g. for CRT. So I imagine that any RSA implementation intending to support such users must use this functionality, which is why I considered it sensible to provide it as a stable, public function the implementors can rely on.

Reply:
My wording wasn't careful enough, I didn't mean that they are not useful in general, of course they are useful, they are very handy whenever the user needs to convert between different key representations. But you were talking about alternative implementations and I was thinking in that context.

What I was suggesting in the first place is, that we use these functions automatically inside the library and don't require the user calling them and therefore being able to remove them from the public interface.

Reply:
@yanesca It seems these functions are a bit special in nature: They are not meant to be used by the end-user of the library, but they are meant to be used by people working on alternative RSA implementations. In this sense, while they don't belong to the public RSA-context related interface that the end-user will be facing, they nonetheless should be public, being supportive functions for alternative implementations. At least that's the idea.

Reply:
@yanesca I've encountered the need to convert between (N,E,D) and (N,E,P,Q,DP,PQ,U) representations before, at the interface between crypto APIs. I'm sure it's going to come up sooner or later as people get a key with only one set of private parameters and eventually need to use it with a hardware implementation that requires the other set.

I do agree that there's no strong reason to make them user-facing though. But they should be available to implementers of hardware/external implementations. This still puts a backward compatibility requirement on us.

Should we move them to a different header `rsa_internal.h`? Or maybe we should only expose `import` and `complete`, and tell alt-implementers to use the standard `mbedtls_rsa_context` structure if they want to use the functions — that's only a very small memory overhead over passing the parameters individually.

Reply:
@gilles-peskine-arm @yanesca Gille's second paragraph is in line with what I intended to say. These functions are special in that they are not user facing but facing alternative implementors, and the latter should be able to rely on stability if they use them. 

I don't quite understand the suggestion in last paragraph, though, because the `deduce_xxx` functions are explicitly meant to be stable across alternative implementations, while `import` and `complete` would be overwritten, _using_ the `deduce_xxx` functions as helpers. Suggesting alt implementors to use the default implementation for the purpose of parameter completion would mean they have to keep a copy of it alongside their own alternative implementation, which doesn't seem sensible to me.

Reply:
@gilles-peskine-arm I feel uncomfortable about moving them into `rsa_internal.h`, but I like the second idea.

@hanno-arm If the alternative implementation is an HSM or a hardware accelerator, then I can't see how they could avoid having copies if they want to use the `deduct_xxx` functions. The `deduct_xxx` functions use `mbedtls_mpi`'s and they have to convert into that format before using them.

Reply:
@yanesca @gilles-peskine-arm Having a separate compilation unit and header for the helper files like `rsa_helper.[c|h]` would be an option, thouh I'm not sure if it really reduces structural complexity. 

@yanesca They surely need a copy of the `deduce_xxx` functions somehow, but I understood @gilles-peskine-arm as suggesting to hide their functionality in `import`-`complete`-`export`-triples; in this case, alternative implementations would need all of these functions from the default implementation together with the default `mbedtls_rsa_context` to access the functionality of `deduce_xxx`, which is a very large and unnecessary overhead.

Reply:
@hanno-arm Now that we have talked it through I think I am understand. I think any alternative solution comes with a trade-off, and therefore I agree that it is best to stay with this one.

Reply:
@yanesca Thank you for the discussions. Regarding the question of whether it's premature to have the helper functions public, we might also discuss that further with @mpg after he had time to look at it. The important point for me is that the functionality is clearly singled out for re-use, but that doesn't necessarily mean it must be public - even though the guarantees for implementors of alternative RSA would be stronger this way. If it's private but clearly documented as being context-independent functionality that's meant to be re-usable for alternative implementations, that might be sufficient support, too.

Reply:
@hanno-arm I agree that Manuel's opinion would be very valuable and we would end up with a better result if we would involve him in this discussion, but I would prefer to be thrifty with team resources and Manuel's time and make a decision on our own. I support the version where we do not make the helper functions public just yet, but I can live with them being exposed too, so it is now up to you.

Reply:
Ok, I'm coming late to a discussion where many points have already been made, so I hope I won't miss or repeat too many of them.

My current understanding is that (1) the use case of importing an (N, E, D) key is covered by `complete()` and I don't see any use case where a simple library user would need to call `deduce_moduli()` directly instead of calling `complete()`, but (2) alt-implementers OTOH might want `deduce_moduli()` to be available (without copy-pasting code), in particular so that they can provide an implementation of `complete()` for normal users.

If that understanding is correct, I think this rules out making `deduce_moduli()` static, and leaves us with two options:
1. Move its declaration to a  new `rsa_internal.h` file (and its implementation either in `rsa.c` outside the `#ifdef ALT` guards, or in `rsa_internal.c`): IMO this is the cleanest option as it clearly reflects its status of being targeted at alt-implementers but not normal users.
2. Leave it in the fully-public API (and keep its implementation outside the `#ifdef ALT` guards). It dislike this option as I'd prefer to keep the public API uncluttered, but I could live with it.

Reply:
@mpg Yes, your understanding matches the intentions.

I also think (1) would be a clean solution.

Reply:
@mpg, @hanno-arm  I like option 1. too.